File: src/main/java/kr/dogfoot/hwpxlib/object/HWPXFile.java
package kr.dogfoot.hwpxlib.object;

import kr.dogfoot.hwpxlib.object.chart.ChartXMLFile;
import kr.dogfoot.hwpxlib.object.common.HWPXObject;
import kr.dogfoot.hwpxlib.object.common.ObjectList;
import kr.dogfoot.hwpxlib.object.common.ObjectType;
import kr.dogfoot.hwpxlib.object.content.context_hpf.ContentHPFFile;
import kr.dogfoot.hwpxlib.object.content.header_xml.HeaderXMLFile;
import kr.dogfoot.hwpxlib.object.content.masterpage_xml.MasterPageXMLFile;
import kr.dogfoot.hwpxlib.object.content.section_xml.SectionXMLFile;
import kr.dogfoot.hwpxlib.object.dochistory.HistoryXMLFile;
import kr.dogfoot.hwpxlib.object.etc.UnparsedXMLFile;
import kr.dogfoot.hwpxlib.object.metainf.ContainerXMLFile;
import kr.dogfoot.hwpxlib.object.metainf.ManifestXMLFile;
import kr.dogfoot.hwpxlib.object.root.SettingsXMLFile;
import kr.dogfoot.hwpxlib.object.root.VersionXMLFile;

import java.util.ArrayList;

/**
 * HWPX 파일 전체를 나타내는 객체
 */
public class HWPXFile extends HWPXObject {
    private final VersionXMLFile versionXMLFile;
    private final ManifestXMLFile manifestXMLFile;
    private final ContainerXMLFile containerXMLFile;
    private final ContentHPFFile contentHPFFile;

    private final HeaderXMLFile headerXMLFile;
    private final ObjectList<MasterPageXMLFile> masterPageXMLFileList;
    private final ObjectList<SectionXMLFile> sectionXMLFileList;
    private final SettingsXMLFile settingsXMLFile;
    private final ObjectList<HistoryXMLFile> historyXMLFileList;
    private final ObjectList<ChartXMLFile> chartXMLFileList;
    private final ArrayList<UnparsedXMLFile> unparsedXMLFileList;

    public HWPXFile() {
        versionXMLFile = new VersionXMLFile();
        manifestXMLFile = new ManifestXMLFile();
        containerXMLFile = new ContainerXMLFile();
        contentHPFFile = new ContentHPFFile();

        headerXMLFile = new HeaderXMLFile();
        masterPageXMLFileList = new ObjectList<MasterPageXMLFile>(MasterPageXMLFile.class);
        sectionXMLFileList = new ObjectList<SectionXMLFile>(SectionXMLFile.class);
        settingsXMLFile = new SettingsXMLFile();
        historyXMLFileList = new ObjectList<HistoryXMLFile>(HistoryXMLFile.class);
        chartXMLFileList = new ObjectList<ChartXMLFile>(ChartXMLFile.class);
        unparsedXMLFileList = new ArrayList<UnparsedXMLFile>();
    }


    @Override
    public ObjectType _objectType() {
        return ObjectType.HWPXFile;
    }

    public VersionXMLFile versionXMLFile() {
        return versionXMLFile;
    }

    public ManifestXMLFile manifestXMLFile() {
        return manifestXMLFile;
    }

    public ContainerXMLFile containerXMLFile() {
        return containerXMLFile;
    }

    public ContentHPFFile contentHPFFile() {
        return contentHPFFile;
    }

    public HeaderXMLFile headerXMLFile() {
        return headerXMLFile;
    }

    public ObjectList<MasterPageXMLFile> masterPageXMLFileList() {
        return masterPageXMLFileList;
    }

    public ObjectList<SectionXMLFile> sectionXMLFileList() {
        return sectionXMLFileList;
    }

    public SettingsXMLFile settingsXMLFile() {
        return settingsXMLFile;
    }

    public ObjectList<HistoryXMLFile> historyXMLFileList() {
        return historyXMLFileList;
    }

    public ObjectList<ChartXMLFile> chartXMLFileList() {
        return chartXMLFileList;
    }

    public void addUnparsedXMLFile(String href, String xml) {
        unparsedXMLFileList.add(new UnparsedXMLFile()
                .hrefAnd(href)
                .xmlAnd(xml));
    }

    public UnparsedXMLFile[] unparsedXMLFiles() {
        return unparsedXMLFileList.toArray(UnparsedXMLFile.ZeroArray);
    }

    public void removeUnparsedXMLFile(UnparsedXMLFile unparsedXMLFile) {
        unparsedXMLFileList.remove(unparsedXMLFile);
    }

    public void removeAllUnparsedXMLFiles() {
        unparsedXMLFileList.clear();
    }

    public HWPXFile clone() {
        HWPXFile cloned = new HWPXFile();
        cloned.copyFrom(this);
        return cloned;
    }

    public void copyFrom(HWPXFile from) {
        versionXMLFile.copyFrom(from.versionXMLFile);
        manifestXMLFile.copyFrom(from.manifestXMLFile);
        containerXMLFile.copyFrom(from.containerXMLFile);
        contentHPFFile.copyFrom(from.contentHPFFile);

        headerXMLFile.copyFrom(from.headerXMLFile);
        for (MasterPageXMLFile masterPageXMLFile : from.masterPageXMLFileList.items()) {
            masterPageXMLFileList.add(masterPageXMLFile.clone());
        }
        for (SectionXMLFile sectionXMLFile : from.sectionXMLFileList.items()) {
            sectionXMLFileList.add(sectionXMLFile.clone());
        }
        settingsXMLFile.copyFrom(from.settingsXMLFile);
        for (HistoryXMLFile historyXMLFile : from.historyXMLFileList.items()) {
            historyXMLFileList.add(historyXMLFile.clone());
        }
        for (ChartXMLFile chartXMLFile : from.chartXMLFileList.items()) {
            chartXMLFileList.add(chartXMLFile.clone());
        }
        for (UnparsedXMLFile unparsedXMLFile : from.unparsedXMLFileList) {
            unparsedXMLFileList.add(unparsedXMLFile.clone());
        }
   }
}

File: js/src/object/HWPXFile.js
import { HWPXObject } from './common/HWPXObject.js';
import { ObjectType } from './common/ObjectType.js';
import { ObjectList } from './common/ObjectList.js';
import { VersionXMLFile } from './root/VersionXMLFile.js';
import { SettingsXMLFile } from './root/SettingsXMLFile.js';
import { ManifestXMLFile } from './metainf/ManifestXMLFile.js';
import { ContainerXMLFile } from './metainf/ContainerXMLFile.js';
import { ContentHPFFile } from './content/context_hpf/ContentHPFFile.js';
import { HeaderXMLFile } from './content/header_xml/HeaderXMLFile.js';
import { SectionXMLFile } from './content/section_xml/SectionXMLFile.js';

/**
 * Unparsed XML file (for files that couldn't be parsed)
 */
export class UnparsedXMLFile {
  constructor() {
    this._href = null;
    this._xml = null;
  }

  href() { return this._href; }
  hrefSet(value) { this._href = value; }
  hrefAnd(value) { this._href = value; return this; }

  xml() { return this._xml; }
  xmlSet(value) { this._xml = value; }
  xmlAnd(value) { this._xml = value; return this; }

  clone() {
    const cloned = new UnparsedXMLFile();
    cloned._href = this._href;
    cloned._xml = this._xml;
    return cloned;
  }
}

/**
 * Chart XML file
 */
export class ChartXMLFile {
  constructor() {
    this._path = null;
    this._data = null;
  }

  path() { return this._path; }
  pathSet(value) { this._path = value; }
  pathAnd(value) { this._path = value; return this; }

  data() { return this._data; }
  dataSet(value) { this._data = value; }
  dataAnd(value) { this._data = value; return this; }

  clone() {
    const cloned = new ChartXMLFile();
    cloned._path = this._path;
    cloned._data = this._data;
    return cloned;
  }
}

/**
 * Main HWPX file object - represents the entire document
 */
export class HWPXFile extends HWPXObject {
  constructor() {
    super();
    this._versionXMLFile = new VersionXMLFile();
    this._manifestXMLFile = new ManifestXMLFile();
    this._containerXMLFile = new ContainerXMLFile();
    this._contentHPFFile = new ContentHPFFile();
    this._headerXMLFile = new HeaderXMLFile();
    this._masterPageXMLFileList = new ObjectList(null); // MasterPageXMLFile
    this._sectionXMLFileList = new ObjectList(SectionXMLFile);
    this._settingsXMLFile = new SettingsXMLFile();
    this._historyXMLFileList = new ObjectList(null); // HistoryXMLFile
    this._chartXMLFileList = new ObjectList(ChartXMLFile);
    this._unparsedXMLFileList = [];
  }

  _objectType() {
    return ObjectType.HWPXFile;
  }

  versionXMLFile() { return this._versionXMLFile; }
  manifestXMLFile() { return this._manifestXMLFile; }
  containerXMLFile() { return this._containerXMLFile; }
  contentHPFFile() { return this._contentHPFFile; }
  headerXMLFile() { return this._headerXMLFile; }
  masterPageXMLFileList() { return this._masterPageXMLFileList; }
  sectionXMLFileList() { return this._sectionXMLFileList; }
  settingsXMLFile() { return this._settingsXMLFile; }
  historyXMLFileList() { return this._historyXMLFileList; }
  chartXMLFileList() { return this._chartXMLFileList; }

  addUnparsedXMLFile(href, xml) {
    this._unparsedXMLFileList.push(
      new UnparsedXMLFile().hrefAnd(href).xmlAnd(xml)
    );
  }

  unparsedXMLFiles() {
    return this._unparsedXMLFileList;
  }

  removeUnparsedXMLFile(file) {
    const index = this._unparsedXMLFileList.indexOf(file);
    if (index !== -1) {
      this._unparsedXMLFileList.splice(index, 1);
    }
  }

  removeAllUnparsedXMLFiles() {
    this._unparsedXMLFileList = [];
  }

  /**
   * Gets all text from all sections
   * @param {string} paraSeparator - Separator between paragraphs
   * @param {string} sectionSeparator - Separator between sections
   * @returns {string}
   */
  getText(paraSeparator = '\n', sectionSeparator = '\n\n') {
    const sections = [];
    for (const section of this._sectionXMLFileList.items()) {
      sections.push(section.getText(paraSeparator));
    }
    return sections.join(sectionSeparator);
  }

  clone() {
    const cloned = new HWPXFile();
    cloned.copyFrom(this);
    return cloned;
  }

  copyFrom(from) {
    this._versionXMLFile.copyFrom(from._versionXMLFile);
    this._manifestXMLFile.copyFrom(from._manifestXMLFile);
    this._containerXMLFile.copyFrom(from._containerXMLFile);
    this._contentHPFFile.copyFrom(from._contentHPFFile);
    this._headerXMLFile.copyFrom(from._headerXMLFile);

    for (const masterPage of from._masterPageXMLFileList.items()) {
      this._masterPageXMLFileList.add(masterPage.clone ? masterPage.clone() : masterPage);
    }

    for (const section of from._sectionXMLFileList.items()) {
      this._sectionXMLFileList.add(section.clone());
    }

    this._settingsXMLFile.copyFrom(from._settingsXMLFile);

    for (const history of from._historyXMLFileList.items()) {
      this._historyXMLFileList.add(history.clone ? history.clone() : history);
    }

    for (const chart of from._chartXMLFileList.items()) {
      this._chartXMLFileList.add(chart.clone());
    }

    for (const unparsed of from._unparsedXMLFileList) {
      this._unparsedXMLFileList.push(unparsed.clone());
    }
  }
}

export default HWPXFile;
